Map/Test/GameTest.cpp

```diff
@@ -0,0 +1,23 @@
+#include "Game.hpp"
+#include "SquareMap.hpp"
+#include <gtest/gtest.h>
+
+using namespace ::testing;
+
+namespace
+{
+    constexpr auto mapSize = 8u;
+    SquareMap map(mapSize);
+}
+
+class GameTest : public Test
+{
+protected:
+    GameTest() : game(map) {}
+    Game game;
+};
+
+TEST_F(GameTest, isBuildingWell)
```

Comment:
brakuej testów

Reply:
Aktualnie funkcja getDirection() nie przyjmuje żadnych argumentów i wymaga uruchomienia w terminalu, żeby pobrać znak od użytkownika. Żeby ją przetestować trzeba przebudować funkcję tak, żeby jako argument przyjmowała znak z klawiatury, czy jest jakiś inny sposób?

Reply:
Właściwie to przebudowanie tej funkcji jest dobrym pomysłem. Uniezalezniamy wtedy naszą klasę od zewnętrznych narzędzi, jakim jest biblioteka graficzna. Wywołania ncruses byłyby tylko w mainie. Ale możesz tez sprobowac w tescie cos takego

 std::istringstream stream("w");
 std::cin.rdbuf(stream.rdbuf());

Kiedyś przy projekcie z ncurses mi to działało. Ale było to dawno i nie jest to eleganckie rozwiazanie :P 

Wytłumaczę o co chodzi. Mamy grę, która wykorzystuje jakiś interfejs graficzny XXX, przez który otrzymujemy dane. Wiemy, że nasza gra ma współpracować z interfejsem XXX, więć uzywamy go we wszystkich klasach (np korzystamy z getch() ). Po roku, szef mrkentingu decyduje, że interfejs XXX jest do dupy, teraz będziemy korzystać z interfejsu YYY (nie ma tu funkcji getch(), tylko jakaś inna). I tutaj pojawia się problem, ponieważ cały nasz projekt korzysta z XXX i jego funckji, dlatego musimy praktycznie przebudować cały system.

Jest też druga, lepsza możliwość, polegająca na izolowaniu się od zewnętrznych narzędzi, np inerfejsów graficznych. Klasy w programie operują tylko na wartościach i funkcjach, które oferuje nam klasyczny C++. Mamy za to jedna klasę, która konwertuje dane z interfejsu. W momencie kiedy zmieniamy interfejs na YYY, wystarczy podmienić tylko jedną klasę, bez rozgrzebywania całego projektu.

Tutaj to średnio ma znaczenie, nie stowrzymy i tak takiej izolacji. Ale dobrze o tym pamiętać na przyszłość :P

Reply:
Dzięki za wyjaśnienie, w tym przypadku faktycznie wystarczy spróbować tego prostszego rozwiązania. Jeszcze mam pytanie o podłączanie ncurses.h w CMakeLists.txt. Próbowałem różnych sposobów, ale ciągle coś jest źle. 
![Zrzut ekranu z 2020-05-02 18-39-56](https://user-images.githubusercontent.com/56999432/80870108-9c1ea280-8ca4-11ea-9aea-3ad2414bf7fc.png)

Reply:
UTeki sie nie budują, bo nie zalinkowales tam ncurses (tak mi sie wydaje )

Reply:
Rzeczywiście, zapomniałem o tym. Teraz wszystko się buduje, ale po włączeniu testów dzieją się dziwne rzeczy. Najpierw nie wyświetla się nic i trzeba nacisnąć dowolny przycisk, aby kontynuować, potem pojawiają się kolejne testy, ale trzeba pomiędzy nimi coś wpisywać. Wygląda na to, że wpisywany tekst jest traktowany jako wartości dla getch(). 
![Zrzut ekranu z 2020-05-04 16-32-10](https://user-images.githubusercontent.com/56999432/80977690-5b01cc00-8e25-11ea-8970-5d1455d82f22.png)
![Zrzut ekranu z 2020-05-04 16-32-21](https://user-images.githubusercontent.com/56999432/80977711-62c17080-8e25-11ea-903a-c0a44993b1f7.png)
![Zrzut ekranu z 2020-05-04 16-32-28](https://user-images.githubusercontent.com/56999432/80977722-66ed8e00-8e25-11ea-8cf8-dee8f98cb6ed.png)
![Zrzut ekranu z 2020-05-04 16-32-30](https://user-images.githubusercontent.com/56999432/80977760-779e0400-8e25-11ea-9eb2-0ea9a14ccb39.png)

Reply:
Mam dwa pomysły, pierwszy to zrobienie "w\n". Jak zadziało to fajnie, ale nie jest to dla nas optymalne rozwiazanie, że trzeba klikać enter po wpisaniu kierunku. 
Drugi mój pomysł to brak ubicia ncurses. Robisz w jednej metodzie init screen, ale nigdzie nie robisz endwin(). Daj znać czy pomogło

Reply:
Tutaj nie widać, ale dałem endwin() tuż przed każdym returnem. Spróbuję użyć "w\n".

Reply:
Niestety to nie rozwiązało problemu. Za każdym razem trzeba ręcznie wpisywać kierunki. Można za to powiedzieć, że sama funkcja działa dobrze, bo wyniki się zgadzają dla ręcznie wpisywanych wartości. Oczywiście wiem, że tak nie może zostać.

Dodanie przed testami
std::istringstream stream("w");
std::cin.rdbuf(stream.rdbuf());
właściwie nie ma żadnego wpływu na testy. Zarówno z tym jak i bez tego wygląda to identycznie.

Jeśli nie ma sposobu na wprowadzanie danych dla getch() innego od ręcznego wpisywania, to może rozdzielę funkcję na getCharacter() i getDirection(), która jako argument przyjmie właśnie ten znak pobrany przez getCharacter(). Umożliwi to testowanie getDirection(), ale wciąż nie będzie można z poziomu testów sprawdzić getCharacter().

Reply:
Ok. Zostaw to jak jest, wykonaleś kawal dobrej pracy, nie będziemy się nad ty meczyć bo się nam konczy czas :P Jak poprawiłes inne rzeczy to mozesz mergowac

Reply:
Jeszcze tylko się upewnię. W przypadku, gdy gracz poda znak inny niż wsad (lub WSAD) lepiej zrobić return getDirection() zamiast break. Prawda?